File: deadlock.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func redRobot(lock1, lock2 *sync.Mutex) {
	for {
		fmt.Println("RED: Acquiring lock 1")
		lock1.Lock()
		fmt.Println("RED: Acquiring lock 2")
		lock2.Lock()

		fmt.Println("RED: locks acquired")
		lock1.Unlock()
		lock2.Unlock()
		fmt.Println("RED: locks released")
		time.Sleep(500 * time.Millisecond)
	}
}

func blueRobot(lock1, lock2 *sync.Mutex) {
	for {
		fmt.Println("BLUE: Acquiring lock 2")
		lock2.Lock()
		fmt.Println("BLUE: Acquiring lock 1")
		lock1.Lock()

		fmt.Println("BLUE: locks acquired")
		lock1.Unlock()
		lock2.Unlock()
		fmt.Println("BLUE: locks released")
		time.Sleep(500 * time.Millisecond)
	}
}

func main() {
	mutex1 := &sync.Mutex{}
	mutex2 := &sync.Mutex{}

	go redRobot(mutex1, mutex2)
	go blueRobot(mutex1, mutex2)

	select {}  // This will keep main from exiting
}

File: deadlock.py
import time
from threading import Lock, Thread

def red_robot(lock_1, lock_2):
    while True:
        print("RED: Acquiring lock 1")
        lock_1.acquire()
        print("RED: Acquiring lock 2")
        lock_2.acquire()

        print("RED: locks acquired")
        lock_1.release()
        lock_2.release()
        print("RED: locks released")
        time.sleep(0.5)

def blue_robot(lock_1, lock_2):
    while True:
        print("BLUE: Acquiring lock 2")
        lock_2.acquire()
        print("BLUE: Acquiring lock 1")
        lock_1.acquire()

        print("BLUE: locks acquired")
        lock_1.release()
        lock_2.release()
        print("BLUE: locks released")
        time.sleep(0.5)


mutex1=Lock()
mutex2=Lock()

red = Thread(target=red_robot, args=(mutex1, mutex2))
blue = Thread(target=blue_robot, args=(mutex1, mutex2))

red.start()
blue.start()